vllm/entrypoints/openai/api_server.py

```diff
@@ -71,6 +76,17 @@ def parse_args():
 route.path_regex = re.compile('^/metrics(?P<path>.*)$')
 app.routes.append(route)
 
+# Siyu added
+@app.middleware("http")
+async def e2e_latency(request: Request, call_next):
```

Comment:
Could you briefly explain what is this `e2e_latency`'s functionality and where it is invoked?

Reply:
> Could you briefly explain what is this `e2e_latency`'s functionality and where it is invoked?

The timing captured by this middleware is essentially the duration of processing on the server side—from the moment the request is handed off to the FastAPI application (right before invoking the endpoint logic or other downstream middleware) until the response is fully generated and ready to be returned. Once the request is processed by the downstream handlers (routes, other middleware, and responses), the middleware resumes control to measure the finish time, calculates the total processing latency, and then logs it.